core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1153,154 +1101,207 @@ class KafkaController(val config: KafkaConfig, zkUtils: ZkUtils, val brokerState
-          // that need to be on this broker
-          if (imbalanceRatio > (config.leaderImbalancePerBrokerPercentage.toDouble / 100)) {
-            topicsNotInPreferredReplica.keys.foreach { topicPartition =>
-              inLock(controllerContext.controllerLock) {
-                // do this check only if the broker is live and there are no partitions being reassigned currently
-                // and preferred replica election is not in progress
-                if (controllerContext.liveBrokerIds.contains(leaderBroker) &&
-                    controllerContext.partitionsBeingReassigned.isEmpty &&
-                    controllerContext.partitionsUndergoingPreferredReplicaElection.isEmpty &&
-                    !deleteTopicManager.isTopicQueuedUpForDeletion(topicPartition.topic) &&
-                    controllerContext.allTopics.contains(topicPartition.topic)) {
-                  onPreferredReplicaElection(Set(topicPartition), true)
-                }
-              }
-            }
+    }
+  }
+
+  case class TopicDeletion(var topicsToBeDeleted: Set[String]) extends ControllerEvent {
+    override def process(): Unit = {
+      if (!isActive) return
+      debug("Delete topics listener fired for topics %s to be deleted".format(topicsToBeDeleted.mkString(",")))
+      val nonExistentTopics = topicsToBeDeleted -- controllerContext.allTopics
+      if (nonExistentTopics.nonEmpty) {
+        warn("Ignoring request to delete non-existing topics " + nonExistentTopics.mkString(","))
+        nonExistentTopics.foreach(topic => zkUtils.deletePathRecursive(getDeleteTopicPath(topic)))
+      }
+      topicsToBeDeleted --= nonExistentTopics
+      if (config.deleteTopicEnable) {
+        if (topicsToBeDeleted.nonEmpty) {
+          info("Starting topic deletion for topics " + topicsToBeDeleted.mkString(","))
+          // mark topic ineligible for deletion if other state changes are in progress
+          topicsToBeDeleted.foreach { topic =>
+            val preferredReplicaElectionInProgress =
+              controllerContext.partitionsUndergoingPreferredReplicaElection.map(_.topic).contains(topic)
+            val partitionReassignmentInProgress =
+              controllerContext.partitionsBeingReassigned.keySet.map(_.topic).contains(topic)
+            if (preferredReplicaElectionInProgress || partitionReassignmentInProgress)
+              topicDeletionManager.markTopicIneligibleForDeletion(Set(topic))
           }
+          // add topic to deletion list
+          topicDeletionManager.enqueueTopicsForDeletion(topicsToBeDeleted)
+        }
+      } else {
+        // If delete topic is disabled remove entries under zookeeper path : /admin/delete_topics
+        for (topic <- topicsToBeDeleted) {
+          info("Removing " + getDeleteTopicPath(topic) + " since delete topic is disabled")
+          zkUtils.zkClient.delete(getDeleteTopicPath(topic))
```

Comment:
Maybe it's nothing but I saw there is a zkUtils method called deletePath. Isn't it better to use that instead of calling the clients method directly?